source/MaterialXGenShader/HwLightHandler.cpp

```diff
@@ -1,25 +1,20 @@
 #include <MaterialXGenShader/HwLightHandler.h>
 #include <MaterialXGenShader/HwShaderGenerator.h>
 
+#include <iterator>
+
 namespace MaterialX
 {
 
-unsigned int HwLightHandler::getLightType(NodePtr node)
+int HwLightHandler::getLightType(NodePtr node) const
```

Comment:
I'm not sure automatic generation of light types is a good idea.. In the old code the client was responsible for setting light type id's, I'm not sure why that was changed?

The problem I see with automatic generation is that once a shader has been generated the client must know which light types id's is used inside the shader, in order to bind lights of the right type. And to know that a client must now keep this HwLightHandler around and query this method. For example, what if a shader source code is cached and not re-generated each time. Also, the type ids are not deterministic since they dependent on the order of adding lights now.

I think the most safe and simple solution is to let the client control the ids. Node ids is something most apps use anyway, and if not it's easy to create a name->id map.

Reply:
Sure we can do this. I don't mind getting rid of this, but we should also get rid of bindLightShaders() since it simply hides the id specification.  I think all we need to pass here is the <lightSource + (unsigned int) type identifier>. The optimization in the method we should still do as mentioned to avoid redundant implementation search.

Reply:
Sounds good! Let's sync offline.